File: website/fe_supplier/app/service/supplierService.ts
import axiosInstance from '../config/axios';
import type { ApiResponse } from './authService';

// ============= SUPPLIER TYPES =============

export interface SupplierProfileUpdateRequest {
  fullName?: string;
  phoneNumber?: string;
  avatarUrl?: string;
  businessAddress?: string;
}

export interface SupplierBusinessUpdateRequest {
  taxCode?: string;
  businessLicense?: string;
  businessLicenseUrl?: string;
  foodSafetyCertificate?: string;
  foodSafetyCertificateUrl?: string;
  supplierNotes?: string;
}

export interface SupplierPendingUpdateResponse {
  updateId: string;
  supplier: {
    userId: string;
    fullName: string;
    businessName: string;
    email: string;
  };
  admin?: {
    userId: string;
    fullName: string;
    email: string;
  } | null;
  taxCode?: string;
  businessLicense?: string;
  businessLicenseUrl?: string;
  foodSafetyCertificate?: string;
  foodSafetyCertificateUrl?: string;
  supplierNotes?: string;
  adminNotes?: string;
  updateStatus: 'PENDING' | 'APPROVED' | 'REJECTED';
  createdAt: string;
  processedAt?: string | null;
}

/**
 * Wallet information nested in SupplierResponse
 * Matches backend SupplierResponse.WalletInfo
 */
export interface WalletInfo {
  walletId: string;
  availableBalance: number;      // Số dư khả dụng (có thể rút)
  pendingBalance: number;         // Số dư đang giữ (đơn hàng đang xử lý)
  totalEarnings: number;          // Tổng thu nhập từ trước đến nay
  totalWithdrawn: number;         // Tổng đã rút
  totalRefunded: number;          // Tổng hoàn trả
  monthlyEarnings: number;        // Thu nhập tháng hiện tại
  currentMonth: string;           // Tháng hiện tại (YYYY-MM)
  status: string;                 // ACTIVE, SUSPENDED, FROZEN
  lastWithdrawalDate: string | null; // Lần rút tiền cuối
  createdAt: string;
  updatedAt: string;
}

/**
 * Basic store information nested in SupplierResponse
 * Matches backend SupplierResponse.StoreBasicInfo
 */
export interface StoreBasicInfo {
  storeId: string;
  storeName: string;
  address: string;
  phoneNumber: string;
  status: string; // PENDING, ACTIVE, SUSPENDED, REJECTED
}

/**
 * Complete supplier response from backend
 * Matches backend SupplierResponse.java exactly
 */
export interface SupplierResponse {
  // User basic info
  userId: string;
  keycloakId: string;
  username: string;
  email: string;
  phoneNumber: string;
  fullName: string;
  avatarUrl: string;
  active: boolean;

  // Business information
  businessName: string;
  businessLicense: string;
  businessLicenseUrl: string;
  foodSafetyCertificate: string;
  foodSafetyCertificateUrl: string;
  taxCode: string;
  businessAddress: string;
  businessType: string; // SUPERMARKET, CONVENIENCE_STORE, GROCERY_STORE, DISTRIBUTOR, RESTAURANT, BAKERY, COFFEE_SHOP, OTHER

  // Commission
  commissionRate: number | null;

  // Status
  status: string; // PENDING_VERIFICATION, PENDING_DOCUMENTS, PENDING_STORE_INFO, PENDING_APPROVAL, ACTIVE, SUSPENDED, REJECTED

  // Wallet information (nested object)
  wallet: WalletInfo | null;

  // Statistics
  totalProducts: number | null;
  totalStores: number | null;
  stores: StoreBasicInfo[] | null;

  // Metadata
  createdAt: string;
  updatedAt: string;
}

// ============= SUPPLIER SERVICE =============

class SupplierService {
  /**
   * Get current supplier profile
   */
  async getCurrentSupplier(): Promise<SupplierResponse> {
    try {
      const response = await axiosInstance.get<ApiResponse<SupplierResponse>>('/suppliers/me');
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Get supplier by ID (admin only)
   */
  async getSupplierById(userId: string): Promise<SupplierResponse> {
    try {
      const response = await axiosInstance.get<ApiResponse<SupplierResponse>>(`/suppliers/${userId}`);
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Update supplier profile
   */
  async updateProfile(request: SupplierProfileUpdateRequest): Promise<SupplierResponse> {
    try {
      const response = await axiosInstance.put<ApiResponse<SupplierResponse>>(
        '/suppliers/me',
        request
      );
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Get all suppliers (admin only)
   */
  async getAllSuppliers(page: number = 0, size: number = 20, status?: string): Promise<any> {
    try {
      const params = new URLSearchParams();
      params.append('page', page.toString());
      params.append('size', size.toString());
      if (status) params.append('status', status);

      const response = await axiosInstance.get<ApiResponse<any>>(
        `/suppliers?${params.toString()}`
      );
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Approve supplier (admin only)
   */
  async approveSupplier(userId: string): Promise<SupplierResponse> {
    try {
      const response = await axiosInstance.patch<ApiResponse<SupplierResponse>>(
        `/suppliers/${userId}/approve`
      );
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Reject supplier (admin only)
   */
  async rejectSupplier(userId: string, reason?: string): Promise<void> {
    try {
      const params = reason ? `?reason=${encodeURIComponent(reason)}` : '';
      await axiosInstance.patch(
        `/suppliers/${userId}/reject${params}`
      );
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Request business info update (requires admin approval)
   */
  async requestBusinessInfoUpdate(request: SupplierBusinessUpdateRequest): Promise<SupplierPendingUpdateResponse> {
    try {
      const response = await axiosInstance.post<ApiResponse<SupplierPendingUpdateResponse>>(
        '/suppliers/me/business-info-update',
        request
      );
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Get my business info update requests
   */
  async getMyBusinessInfoUpdates(status?: string, page: number = 0, size: number = 20): Promise<any> {
    try {
      const params = new URLSearchParams();
      params.append('page', page.toString());
      params.append('size', size.toString());
      if (status) params.append('status', status);

      const response = await axiosInstance.get<ApiResponse<any>>(
        `/suppliers/me/business-info-updates?${params.toString()}`
      );
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Get business info update by ID
   */
  async getBusinessInfoUpdateById(updateId: string): Promise<SupplierPendingUpdateResponse> {
    try {
      const response = await axiosInstance.get<ApiResponse<SupplierPendingUpdateResponse>>(
        `/suppliers/business-info-updates/${updateId}`
      );
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Pause operations (supplier self-pause)
   */
  async pauseOperations(reason?: string): Promise<SupplierResponse> {
    try {
      const params = reason ? `?reason=${encodeURIComponent(reason)}` : '';
      const response = await axiosInstance.patch<ApiResponse<SupplierResponse>>(
        `/suppliers/me/pause${params}`
      );
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Resume operations (resume from PAUSE)
   */
  async resumeOperations(): Promise<SupplierResponse> {
    try {
      const response = await axiosInstance.patch<ApiResponse<SupplierResponse>>(
        '/suppliers/me/resume'
      );
      return response.data.data;
    } catch (error: any) {
      throw this.handleError(error);
    }
  }

  /**
   * Handle API errors
   */
  private handleError(error: any): Error {
    if (error.response?.data?.message) {
      return new Error(error.response.data.message);
    }
    return new Error(error.message || 'An unexpected error occurred');
  }
}

// Export singleton instance
export default new SupplierService();

File: backend/src/main/java/com/example/backend/service/SupplierService.java
package com.example.backend.service;

import com.example.backend.dto.request.*;
import com.example.backend.dto.response.RegisterResponse;
import com.example.backend.dto.response.SupplierResponse;
import com.example.backend.entity.enums.SupplierStatus;

/**
 * Service for Supplier management with 4-step registration flow
 */
public interface SupplierService {

    // ========== 4-STEP REGISTRATION FLOW ==========

    /**
     * Step 1: Register supplier account (username, email, password, phone, fullName)
     * Status: PENDING_VERIFICATION
     * @param request Step 1 request
     * @return Registration response with supplierId
     */
    RegisterResponse registerStep1(SupplierRegisterStep1Request request);

    /**
     * Step 2: Verify email with OTP
     * Status: PENDING_VERIFICATION -> PENDING_DOCUMENTS
     * @param supplierId Supplier ID from step 1 (String UUID)
     * @param request Step 2 request with OTP
     * @return Success message
     */
    String verifyEmailStep2(String supplierId, SupplierRegisterStep2Request request);

    /**
     * Resend OTP for email verification
     * @param supplierId Supplier ID (String UUID)
     * @return Success message
     */
    String resendOtp(String supplierId);

    /**
     * Step 3: Upload business documents (businessLicense, foodSafetyCertificate, avatar)
     * Status: PENDING_DOCUMENTS -> PENDING_STORE_INFO
     * @param supplierId Supplier ID from step 1 (String UUID)
     * @param request Step 3 request with document URLs
     * @return Success message
     */
    String uploadDocumentsStep3(String supplierId, SupplierRegisterStep3Request request);

    /**
     * Step 4: Submit store information and business details
     * Status: PENDING_STORE_INFO -> PENDING_APPROVAL (waiting for admin approval)
     * @param supplierId Supplier ID from step 1 (String UUID)
     * @param request Step 4 request with business info and store details
     * @return Success message
     */
    String submitStoreInfoStep4(String supplierId, SupplierRegisterStep4Request request);

    // ========== PROFILE MANAGEMENT ==========

    /**
     * Get supplier information by keycloakId
     * @param keycloakId Keycloak user ID
     * @return Supplier response
     */
    SupplierResponse getSupplierInfo(String keycloakId);

    /**
     * Get public supplier information (hide sensitive data)
     * @param supplierId Supplier ID (String UUID)
     * @return Public supplier response
     */
    SupplierResponse getPublicSupplierInfo(String supplierId);

    /**
     * Get supplier information by userId
     * @param userId User ID (String UUID)
     * @return Supplier response
     */
    SupplierResponse getSupplierById(String userId);

    /**
     * Update supplier profile (name, phone, avatar, businessAddress)
     * @param keycloakId Keycloak user ID
     * @param request Update request
     * @return Updated supplier response
     */
    SupplierResponse updateProfile(String keycloakId, SupplierProfileUpdateRequest request);

    // ========== ADMIN OPERATIONS ==========

    /**
     * Get all suppliers with pagination, search, and filtering
     * @param page Page number (0-indexed)
     * @param size Page size
     * @param status Filter by status (optional)
     * @param search Search by name, email, or business name (optional)
     * @param sortBy Sort field (optional, default: createdAt)
     * @param sortDirection Sort direction (ASC/DESC, default: DESC)
     * @return Page of suppliers
     */
    org.springframework.data.domain.Page<SupplierResponse> getAllSuppliers(
            int page,
            int size,
            SupplierStatus status,
            String search,
            String sortBy,
            String sortDirection
    );

    /**
     * Approve supplier (set status to ACTIVE and send email notification)
     * @param userId User ID (String UUID)
     * @param approvalNote Optional note from admin
     * @return Updated supplier response
     */
    SupplierResponse approveSupplier(String userId, String approvalNote);

    /**
     * Reject supplier (set status to REJECTED and send email notification)
     * @param userId User ID (String UUID)
     * @param rejectionReason Reason for rejection
     * @return Updated supplier response
     */
    SupplierResponse rejectSupplier(String userId, String rejectionReason);

    /**
     * Update supplier status (by admin)
     * @param userId User ID (String UUID)
     * @param status New status
     * @return Updated supplier response
     */
    SupplierResponse updateStatus(String userId, SupplierStatus status);

    /**
     * Update supplier commission rate (by admin)
     * @param userId User ID (String UUID)
     * @param request Commission update request
     * @return Updated supplier response
     */
    SupplierResponse updateCommissionRate(String userId, SupplierCommissionUpdateRequest request);

    /**
     * Activate/Deactivate supplier account (by admin)
     * @param userId User ID (String UUID)
     * @param active Active status
     * @return Updated supplier response
     */
    SupplierResponse setActive(String userId, boolean active);
}
